### ISSUE TITLE
Improved decoupling between solver front-end and back-end

### ISSUE THREAD

**Issue by alaurenzi** [robotology/OpenSoT#105]
Right now there seems to be a bit of confusion about what is a solver, what is
its front end and its back end. I think that a rational organization could be the following:
- A *concrete* class that transforms a vector of *tasks* + a global *constraint* into a set of
 standard **linear least squares problems**. This class should also handle enforcement of
 priorities. To me, this is the meaning of **front end**.
- An *abstract* interface to implement back-ends, i.e. an object whose role is to actually **solve**
a **linear least squares problem**. Instances of this class should be *loaded at runtime* by the front end.

Right now, the class `Solver` is the front-end, but is just an interface. As a consequence, all the QP matrices generation process is **replicated**, for instance, between the iHQP and the eHQP implementations, which is bad. It makes it really difficult to implement another solver. 

The back-end, on the contrary, is right now a concrete class implemented with qpOASES. For this reason, OpenSot has to *link against qpOASES*, which again is bad. The user should be able to plug his/her own back-end inside the library **without having to modify the source code** and **without recompiling OpenSoT**.

Let us discuss about this @EnricoMingo

**Comment by EnricoMingo** [comment]
I agree BUT, in my opinion we should have an interface for bot the **front-end** and **back-end**. This for different reasons: first of all the algorithms of iHQP and eHQP can have slightly different implementations and both are **front-ends**. Furthermore both could uses different **back-ends**. 
Second, we could have some special cases, see HCOD, in which, in the contrary,   **front-end** and  **back-end** are mixed, we should then be able to handle these special cases. 

In general, considering our two solvers, we should have 2 **front-ends** (iHQP and eHQP) and 2 **back-ends** (qpOASES and Pseudo-Inverse Eigen-based).

**Comment by EnricoMingo** [comment]
I have just open a new branch called ```solver_refactor_backend``` which I will use these days to do the refactor of the ***backend*** as in the previous comments. Maybe we can use the refactor also to implement another ***backend*** based on the  new qp solver ***osqp***.

**Comment by alaurenzi** [comment]
I gave a look to the source code and looks good! I will try it as soon as possible!

**Comment by EnricoMingo** [comment]
We are working in these issue in ```solver_refactor_backend``` branch.
